File: tools/empty_sections.rb
# ruby empty_sections.rb -p ../schema

require 'nokogiri'
require 'optparse'

VERSION = '1.0.0'.freeze

def create_path(path)
  path.tr('\\', '/')
end

# implement commandline options
options = {:path => nil}

parser = OptionParser.new do |opts|
  opts.banner = 'Usage: empty_sections.rb'

  opts.on('-p', '--path path', 'Path to schema') do |path|
    options[:path] = path
  end

  opts.on('-h', '--help', 'Displays help') do
    puts opts
    exit
  end

  opts.on_tail('--version', 'Show program version') do
    puts VERSION
    exit
  end
end

parser.parse!

if options[:path].nil?
  print 'Enter path to schema: '
  options[:path] = STDIN.gets.chomp
end

# start
project_path = create_path(options[:path])
@error = 0

Dir.glob("#{project_path}/**/{CF1R,CF2R,CF3R,NRCV}/*.xsd").map do |schema|
  doc = Nokogiri::XML(File.open(schema))
  doc.xpath('/xsd:schema/xsd:complexType/xsd:sequence/xsd:element').map do |section|
    letter = section.attribute('name').to_s.split('_').last
    if section.xpath('xsd:complexType/xsd:sequence/*').size.zero?
      @error = 1
      puts "#{File.basename(schema)} has an empty Section #{letter}\n"
    end
  end
end

exit 1 unless @error.zero?

File: tools/empty_sections.py
"""Python check for empty schema sections.
cd tools
python3 empty_sections.py ../schema
"""

import glob
import os
import sys
from lxml import etree


# start
error = 0  # pylint:disable=C0103

for filename in glob.glob(os.path.join(sys.argv[1], '**/*.xsd'), recursive=True):
    # TODO: fix error # pylint:disable=W0511
    with open(filename, 'r') as schema:
        data = schema.read()
        tree = etree.fromstring(str.encode(data))  # pylint:disable=I1101
        section = tree.xpath('/xsd:schema/xsd:complexType/xsd:sequence/xsd:element',
                             namespaces={'xsd': 'http://www.w3.org/2001/XMLSchema'})
        try:
            letter = section[0].xpath('@name')[0].split('_')[1]
            if not section[0].xpath('xsd:complexType/xsd:sequence/*',
                                    namespaces={'xsd': 'http://www.w3.org/2001/XMLSchema'}):
                error = 1
                print('{} has an empty Section {}'.format(os.path.basename(filename), letter))
        except IndexError:
            pass

if error == 1:
    exit(1)
